### PR TITLE
[YW-307] fix(engine-server): flaky atomic-ingest pending-query race

### PR DESCRIPTION
**Pull request by youhaowei**
## Summary

Fixes the intermittent `"Attempting to execute an unsuccessful or closed pending query result"` flake in `native-engine.test.ts` — the "atomic ingest" test that blocked unrelated PRs on every CI re-run.

## Root cause

In `registerArrowTable`'s catch block, after `appender.closeSync()` on a failed appender, the cleanup `DROP TABLE` was issued on the **same connection (`conn`)**. On Linux, `duckdb_appender_close()` on a failed appender marks the connection with a pending-result error. The next `duckdb_query()` on that connection then fails with the reported error — **emitted as an unhandled NAPI-layer rejection that bypasses the surrounding JS `try/catch`**, so the error surfaced and killed the test.

## Fix

Route the catch-block cleanup `DROP TABLE` through a **fresh `DuckDBConnection`** from the same `DuckDBInstance`. The fresh connection is not affected by the appender's error state.

**Important caveat (verified empirically):** DuckDB TEMP tables are connection-local — the fresh connection cannot see the staging table, so the DROP is a no-op (passes only via `IF EXISTS`). The staging table persists until `dispose()` clears the main connection. The fix's value is **crash prevention** (avoiding the unhandled NAPI rejection), not early cleanup. Comments updated throughout to reflect this accurately.

## 50× loop result

```
=== 50× RESULT: 50/50 passed, 0/50 failed ===
```

Run under `npx vitest run` (77 tests × 50 iterations), same runner as CI. Note: the Linux-specific NAPI taint cannot be reproduced on macOS/arm64; the 50× confirms no regression on the success path and cross-platform API correctness of the fresh-connection pattern.

## Changes

- **`native-engine.ts`**: Catch block now uses `this.instance!.connect()` for the cleanup, disconnects in `finally`, swallows connect/disconnect failures. Block comment updated to accurately describe success-vs-failure-path cleanup semantics and TEMP scope.
- **`native-engine.test.ts`**: New `describe` block at the `DuckDBInstance` layer proves a fresh connection can issue `DROP` after appender error without throwing. Separately from the existing "atomic ingest" contract test (which continues to verify the observable: prior table row count is intact after a failed registration).

## Screenshots

No UI change — `engine-server` internals only.

Tracked internally as YW-307.

<details open><summary><h3>Greptile Summary</h3></summary>

Fixes the intermittent \"pending-result\" crash in `registerArrowTable` by disconnecting the Linux-tainted connection in the catch block and replacing it with a fresh one from the same `DuckDBInstance`, rather than issuing a `DROP TABLE` on the already-broken connection.

- **`native-engine.ts`**: The catch block now calls `disconnectSync()` on `this.connection`, nulls it, and reconnects — letting DuckDB implicitly drop the partial staging TEMP table when the old connection closes. The success path is unchanged and still uses the original `conn` reference, which is safe because the reconnect only happens on the throw path.
- **`native-engine.test.ts`**: Two new `describe` blocks pin the fix: one at the raw `DuckDBInstance` layer (disconnect + fresh-connect survives appender error) and one at the engine contract level (subsequent `query()` and `registerArrowTable()` succeed after a failed ingest).
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the disconnect-and-reconnect approach is mechanically correct and the 50x loop confirms no regressions.

The catch block always throws so the stale conn local is never reached on the error path; non-TEMP tables survive the connection swap because they are instance-scoped; _registeredTables remains accurate throughout.

The reconnect-failure branch in native-engine.ts (lines 260-270) could benefit from clearing initPromise to keep initialize() usable as a recovery path.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/engine-server/src/native-engine.ts | Catch block now disconnects and replaces the tainted connection instead of issuing a DROP on the same tainted connection; a reconnect-failure edge case leaves initPromise stale and initialize() unable to recover. |
| packages/engine-server/src/native-engine.test.ts | Two new describe blocks added: a DuckDBInstance-layer mechanism test and an engine-level contract test for reuse after failure; both are well-structured with clear preconditions and accurate comments about platform-specific behavior. |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Caller
    participant Engine as NativeDuckDBEngine
    participant Conn as this.connection
    participant Instance as DuckDBInstance

    Caller->>Engine: registerArrowTable(df, buf)
    Engine->>Conn: run(CREATE TEMP TABLE __staging_...)
    Engine->>Conn: createAppender(__staging_...)
    Engine->>Conn: appendArrowValue x N rows
    Note over Conn: append throws
    Engine->>Conn: appender.closeSync() swallow
    Engine->>Conn: disconnectSync() TEMP table auto-dropped
    Engine->>Engine: "this.connection = null"
    Engine->>Instance: connect() fresh connection
    Engine->>Engine: "this.connection = freshConn"
    Engine-->>Caller: rejects with original error
    Caller->>Engine: query(SELECT ...)
    Engine->>Instance: runAndReadAll via freshConn
    Instance-->>Caller: QueryResult success
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Caller
    participant Engine as NativeDuckDBEngine
    participant Conn as this.connection
    participant Instance as DuckDBInstance

    Caller->>Engine: registerArrowTable(df, buf)
    Engine->>Conn: run(CREATE TEMP TABLE __staging_...)
    Engine->>Conn: createAppender(__staging_...)
    Engine->>Conn: appendArrowValue x N rows
    Note over Conn: append throws
    Engine->>Conn: appender.closeSync() swallow
    Engine->>Conn: disconnectSync() TEMP table auto-dropped
    Engine->>Engine: "this.connection = null"
    Engine->>Instance: connect() fresh connection
    Engine->>Engine: "this.connection = freshConn"
    Engine-->>Caller: rejects with original error
    Caller->>Engine: query(SELECT ...)
    Engine->>Instance: runAndReadAll via freshConn
    Instance-->>Caller: QueryResult success
```

</a>
</details>

<sub>Reviews (4): Last reviewed commit: ["test(engine-server): use Uint64 overflow..."](https://github.com/youhaowei/dashframe/commit/632c256c8a3193c84217a43a6bbdcdb751eb8076) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=40261388)</sub>

### CODE REVIEW

**Comment by youhaowei** [comment]
Addressing CodeRabbit's two findings:

**Finding 1 (assert precondition):** Fixed in the follow-up commit (0cec87bc). Added `expect(() => appender.flushSync()).toThrow()` to the mechanism test so it fails loudly if the appender doesn't actually enter error state.

**Finding 2 (primary connection taint):** The concern is whether `this.connection` (the engine's primary conn) remains tainted after a failed appender, making subsequent `engine.query()` calls fail. After investigation:

1. The taint is **one-shot** — it is triggered specifically when `conn.run()` is called on a connection where the previous appender `closeSync()` emitted an unhandled NAPI rejection. The original code triggered this by calling `conn.run(DROP TABLE)` on the tainted conn. The fix routes that DROP to a fresh connection, so the NAPI rejection never fires.

2. The primary `conn` itself recovers after `appender.closeSync()` — empirically verified on macOS: after the full catch-block sequence (appender error → closeSync → fresh-conn DROP), `conn.run('SELECT 1')` succeeds. The ticket's reported flake was only in the cleanup DROP, not in subsequent engine queries.

3. The existing **"atomic ingest" test** (line 242-271) already covers this: it calls `engine.query('SELECT COUNT...')` on the same `NativeDuckDBEngine` instance after `rejects.toThrow()`. If `conn` were permanently tainted, that query would fail — but the test passes reliably (50/50 in the loop, green in CI).

No connection reset needed. The fix (fresh connection for the catch-block DROP) is sufficient and minimal.

**Comment by youhaowei** [comment]
Correction to my previous comment (4816216648): I was wrong about 'no connection reset needed.'

**The actual issue:** `conn` in `registerArrowTable` is `this.connection` — the PERSISTENT connection reused for the whole session (`conn()` getter at line 123). Routing the cleanup DROP through a fresh connection avoids the immediate crash in the catch block, but `this.connection` remains tainted. The next `query()`, `queryArrow()`, or `registerArrowTable()` call reuses the same poisoned handle → the flake relocates to the next operation, not killed.

**Additional note on the atomic-ingest test:** the test uses `tableFromIPC([0xff,0xfe,0x00,0x01])` which returns an empty table on macOS, throwing at the `fields.length === 0` guard (line 197) *before any appender exists*. So that test path never exercises the appender failure at all — my claim that it covers post-failure query health was incorrect.

**The correct fix:** replace `this.connection` in the catch block by disconnecting it and reconnecting from `this.instance`. A fresh connection is clean by construction. This also properly handles the TEMP staging table — when the old connection closes, DuckDB automatically drops its connection-local TEMP tables, finally achieving the early cleanup the no-op DROP couldn't. Commit incoming.

**Comment by youhaowei** [comment]
Commit 23de814f addresses both CodeRabbit findings completely.

**What changed:**
- The catch block now disconnects `this.connection` and reconnects a fresh connection from the same instance, instead of issuing a cleanup DROP on a separate fresh connection. This kills the flake at the source: the persistent connection (`this.connection`, returned by `conn()` and reused for all `query()`/`queryArrow()`/`registerArrowTable()` calls) is replaced with a clean one, so no subsequent operation inherits the tainted state.
- A bonus: when the old tainted connection closes, DuckDB automatically drops all its connection-local TEMP tables — the partial staging table is cleaned up for free without a separate DROP.

**Tests added (engine-level regression, not just mechanism):**
- *DuckDB connection-reset mechanism*: verifies the disconnect + reconnect sequence at the `DuckDBInstance` layer. After appender error → `closeSync` → `disconnectSync` → fresh connect, the new connection can execute queries AND the staging TEMP table is confirmed gone (queried via `duckdb_tables()`).
- *Engine reuse after failure*: verifies `query()` and `registerArrowTable()` both succeed on the SAME `NativeDuckDBEngine` instance after a failure that goes through the actual appender path (overflow `TimestampMillisecond` value → appender throws inside the loop). This is the engine-level contract test CodeRabbit asked for.

File: packages/engine-server/src/native-engine.test.ts [hunk 1]
```diff
@@ -353,3 +354,67 @@ describe("NativeDuckDBEngine — real native DuckDB (Stage 3)", () => {
...
+      try {
+        appender.flushSync();
+      } catch {
+        /* expected: incomplete row */
+      }
+      try {
+        appender.closeSync();
+      } catch {
+        /* expected: appender in error state */
+      }
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine-server/src/native-engine.test.ts]: Fixed in commits 0cec87bc + 23de814f. The mechanism test now asserts `expect(() => appender.flushSync()).toThrow()` (precondition) and the assertions check observable state (table gone, fresh conn returns 1) — not just `.resolves.toBeDefined()`. The incomplete-row approach reliably reproduces the Linux taint on any platform and is tested separately from the engine-level reuse regression.

File: packages/engine-server/src/native-engine.ts [hunk 2]
```diff
@@ -242,10 +244,31 @@ export class NativeDuckDBEngine implements QueryEngine {
...
+        try {
+          await cleanupConn.run(
+            `DROP TABLE IF EXISTS ${quoteIdent(stagingName)}`,
+          );
+        } finally {
+          cleanupConn.disconnectSync();
+        }
       } catch {
-        // best-effort cleanup
+        // best-effort — swallow any connect-or-disconnect failure
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine-server/src/native-engine.ts]: Added in commit 23de814f, refined in 632c256c. The new 'NativeDuckDBEngine — reuse after registerArrowTable failure' describe block tests the exact reuse path: (1) register a good table, (2) trigger a failure INSIDE the append loop via a Uint64 value of 2^63 (appendBigInt throws 'bigint out of int64 range' after the appender is open), (3) assert query() on the pre-existing table still works, (4) assert a fresh registerArrowTable() succeeds on the same engine instance. The assertion uses .rejects.toThrow('bigint out of int64 range') to verify the catch block is actually exercised, not a pre-block decode error.

File: packages/engine-server/src/native-engine.ts [hunk 3]
```diff
@@ -242,10 +244,31 @@ export class NativeDuckDBEngine implements QueryEngine {
...
+      // The table drops automatically when `conn` is eventually disconnected
+      // (dispose()). Early cleanup is therefore not achieved here; the goal is
+      // solely to avoid the unhandled NAPI rejection.
       try {
-        await conn.run(`DROP TABLE IF EXISTS ${quoteIdent(stagingName)}`);
+        const cleanupConn = await this.instance!.connect();
+        try {
+          await cleanupConn.run(
+            `DROP TABLE IF EXISTS ${quoteIdent(stagingName)}`,
+          );
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine-server/src/native-engine.ts]: Resolved in commit 23de814f. The fresh-connection DROP is gone entirely. Instead, the catch block calls this.connection.disconnectSync() + await this.instance.connect() to reset the persistent connection. When the old connection closes, DuckDB automatically drops all its connection-local TEMP tables (including any partial staging table) — so no explicit DROP is needed. A fresh connection that's never seen the staging table can't accidentally drop a persistent table with the same name.

File: packages/engine-server/src/native-engine.ts [hunk 4]
```diff
@@ -242,10 +244,31 @@ export class NativeDuckDBEngine implements QueryEngine {
...
+      // from the same instance is not affected by the appender's error state.
+      //
+      // Note: staging tables are TEMP and connection-local — a fresh connection
+      // cannot see them, so the DROP is a no-op (passes only via IF EXISTS).
+      // The table drops automatically when `conn` is eventually disconnected
+      // (dispose()). Early cleanup is therefore not achieved here; the goal is
+      // solely to avoid the unhandled NAPI rejection.
       try {
-        await conn.run(`DROP TABLE IF EXISTS ${quoteIdent(stagingName)}`);
+        const cleanupConn = await this.instance!.connect();
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine-server/src/native-engine.ts]: This is the core fix, implemented in commit 23de814f. The catch block now calls this.connection.disconnectSync() then this.connection = await this.instance.connect(). Since the instance (and all non-TEMP tables in it) is unaffected by the appender error, the new connection is clean and all subsequent operations (query, queryArrow, registerArrowTable) go through an untainted handle. The 'engine remains usable after a failed ingest' regression test (632c256c) verifies the SAME engine instance can query pre-existing tables and register new ones after the failure — this is the end-to-end contract being protected.
